tls/s2n_client_key_exchange.c

```diff
@@ -203,17 +212,21 @@ int s2n_ecdhe_client_key_send(struct s2n_connection *conn, struct s2n_blob *shar
 
 int s2n_rsa_client_key_send(struct s2n_connection *conn, struct s2n_blob *shared_key)
 {
-    uint8_t client_protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN];
-    client_protocol_version[0] = conn->client_protocol_version / 10;
-    client_protocol_version[1] = conn->client_protocol_version % 10;
+    uint8_t client_hello_protocol_version[S2N_TLS_PROTOCOL_VERSION_LEN];
+    uint8_t legacy_client_hello_protocol_version = get_client_hello_protocol_version(conn);
+    client_hello_protocol_version[0] = legacy_client_hello_protocol_version / 10;
+    client_hello_protocol_version[1] = legacy_client_hello_protocol_version % 10;
 
     shared_key->data = conn->secure.rsa_premaster_secret;
     shared_key->size = S2N_TLS_SECRET_LEN;
 
     GUARD(s2n_get_private_random_data(shared_key));
 
-    /* Over-write the first two bytes with the client protocol version, per RFC2246 7.4.7.1 */
-    memcpy_check(conn->secure.rsa_premaster_secret, client_protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN);
+    /* Over-write the first two bytes with the client hello version, per RFC2246/RFC4346/RFC5246 7.4.7.1.
+     * The latest version supported by client (as seen from the the client hello version) are <= TLS1.2
+     * for all clients, because TLS 1.3 clients freezes the TLS1.2 legacy version in client hello.
+     */
+    memcpy_check(conn->secure.rsa_premaster_secret, client_hello_protocol_version, S2N_TLS_PROTOCOL_VERSION_LEN);
```

Comment:
nit: any reason not to use the same way of doing this we do in the `s2n_rsa_client_key_recv` function?

generally, there is a good chance to refactor here, as they both do the following:

1. `get_client_hello_protocol_version`
2. decompose to wire format
3. `s2n_get_private_random_data`
4. Over-write the first two bytes with the client hello version, per RFC2246/RFC4346/RFC5246 7.4.7.1

Reply:
a refactor sounds like a good suggestion but i would prefer to follow up with a separate task for this :)